h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java

```diff
@@ -65,64 +100,127 @@ public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
+        double x = c.isNA(r) ? _noMatch : in(_matchesIndexes, _sortedValues, _mapping, c.atStr(bs, r).toString(), _noMatch, _startIndex);
         nc.addNum(x);
       }
     }
   }
 
   private static class CatMatchTask extends MRTask<CatMatchTask> {
-    String[] _values;
+    String[] _sortedValues;
+    int[] _firstMatchRow;
     double _noMatch;
-    CatMatchTask(String[] values, double noMatch) {
-      _values = values;
+    int _startIndex;
+    IcedHashMap<String, Integer> _mapping;
+    IcedHashMap<String, Integer> _matchesIndexes;
+    
+ 
+    CatMatchTask(String[] values, double noMatch, int startIndex) {
+      _mapping = new IcedHashMap<>();
+      for (int i = 0; i < values.length; i++) {
+        String value = values[i];
+        if (!_mapping.containsKey(value)) _mapping.put(values[i], i);
+      }
+      _sortedValues = values.clone();
+      Arrays.sort(_sortedValues);
       _noMatch = noMatch;
+      _startIndex = startIndex;
+      _firstMatchRow = new int[values.length];
+      _matchesIndexes = new IcedHashMap<>();
     }
+
     @Override
     public void map(Chunk c, NewChunk nc) {
       String[] domain = c.vec().domain();
       int rows = c._len;
       for (int r = 0; r < rows; r++) {
-        double x = c.isNA(r) ? _noMatch : in(_values, domain[(int) c.at8(r)], _noMatch);
+        double x = c.isNA(r) ? _noMatch : in(_matchesIndexes, _sortedValues, _mapping, domain[(int) c.at8(r)], _noMatch, _startIndex);
         nc.addNum(x);
       }
     }
   }
 
   private static class NumMatchTask extends MRTask<CatMatchTask> {
-    double[] _values;
+    double[] _sortedValues;
     double _noMatch;
-    NumMatchTask(double[] values, double noMatch) {
-      _values = values;
+    int _startIndex;
+    IcedHashMap<Double, Integer> _mapping;
+    IcedHashMap<Double, Integer> _matchesIndexes;
+
+    NumMatchTask(double[] values, double noMatch, int startIndex) {
+      _mapping = new IcedHashMap<>();
+      for (int i = 0; i < values.length; i++) {
+        double value = values[i];
+        if (!_mapping.containsKey(value)) _mapping.put(values[i], i);
+      }
+      _sortedValues = values.clone();
+      Arrays.sort(_sortedValues);
       _noMatch = noMatch;
+      _startIndex = startIndex;
+      _matchesIndexes = new IcedHashMap<>();
     }
+ 
     @Override
     public void map(Chunk c, NewChunk nc) {
       int rows = c._len;
       for (int r = 0; r < rows; r++) {
-        double x = c.isNA(r) ? _noMatch : in(_values, c.atd(r), _noMatch);
+        double x = c.isNA(r) ? _noMatch : in(_matchesIndexes, _sortedValues, _mapping, c.atd(r), _noMatch, _startIndex);
         nc.addNum(x);
       }
     }
   }
 
-  private static double in(String[] matches, String s, double nomatch) {
-    return Arrays.binarySearch(matches, s) >= 0 ? 1 : nomatch;
+  private static double in(IcedHashMap<String, Integer> matchesIndexes, String[] sortedMatches, IcedHashMap<String, Integer> mapping, String s, double noMatch, int startIndex) {
```

Comment:
nitpick: better use `Map` interface in signatures: it's also shorter/less noisy :)